libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -1525,145 +1527,160 @@ std::vector<std::weak_ptr<core::Runnable>> BeaconSetupService::GetWeakRunnables(
-  {
-    expected_dkg_time_s = 27229;
-  }
-  if (cabinet_size < 90)
-  {
-    expected_dkg_time_s = 1304;
-  }
-  if (cabinet_size < 60)
-  {
-    expected_dkg_time_s = 305;
-  }
-  if (cabinet_size < 51)
-  {
-    expected_dkg_time_s = 250;
-  }
-  if (cabinet_size < 30)
-  {
-    expected_dkg_time_s = 100;
-  }
-  if (cabinet_size < 10)
-  {
-    expected_dkg_time_s = 30;
-  }
-  if (cabinet_size < 8)
-  {
-    expected_dkg_time_s = 10;
-  }
+  // Empirical map of cabinet size to expected time. Final element is max to guarantee lower bound
+  // isn't end()
+  // clang-format off
+  static /*constexpr*/ std::map<uint64_t, uint64_t> dkg_time_to_cabinet_size_map
+  {{8, 10},
+   {10, 30},
+   {30, 100},
+   {51, 250},
+   {60, 305},
+   {90, 1304},
+   {200, 27229},
+   {std::numeric_limits<uint64_t>::max(), 27229} };
+  // clang-format on
 
-  // This assigns a time in seconds to each slot. Dividing by 100 is an arbitrary way to select the
-  // time for a single slot from the total DKG time and we could equally just set the time per slot
-  // directly.
-  return static_cast<double>(expected_dkg_time_s) / 100.;
+  // Note it is assumed the total dkg time exceeds 1s * DKG states
+  return (dkg_time_to_cabinet_size_map.lower_bound(cabinet_size))->second;
 }
 
-void BeaconSetupService::SetTimeBySlots(BeaconSetupService::State state, uint64_t &time_slots_total,
-                                        uint64_t &time_slot_for_state)
+/**
+ * Given that we are entering State state, with a known starting point in time,
+ * set the deadline for this state to complete
+ *
+ */
+void BeaconSetupService::SetDeadlineForState(BeaconSetupService::State const &state)
 {
-  assert(time_slot_map_.find(state) != time_slot_map_.end());
-  time_slot_for_state = time_slot_map_.at(state);
+  auto it = time_slot_map_.find(state);
 
-  while (state != BeaconSetupService::State::RESET)
+  if (it == time_slot_map_.end())
   {
-    time_slots_total += time_slot_map_.at(state);
-    state = BeaconSetupService::State(static_cast<uint8_t>(state) - 1);
+    FETCH_LOG_ERROR(LOGGING_NAME, NodeString(),
+                    "Attempt to set the time for a state that has no associated time!");
+    return;
   }
+
+  // Need to increment since for_each isn't inclusive
+  ++it;
+
+  uint64_t time_slots_to_end = 0;
+
+  // Walk through the map adding time slots, including the initial state
+  using MapPair = typename decltype(time_slot_map_)::value_type;
+  std::for_each(time_slot_map_.begin(), it, [&time_slots_to_end](MapPair const &cabinet_time_pair) {
+    time_slots_to_end += cabinet_time_pair.second;
+  });
+
+  assert(expected_dkg_timespan_ != 0 && time_slots_in_dkg_ != 0);
+  assert(time_slots_to_end <= time_slots_in_dkg_);
+
+  // Note: fine to do floor arithmetic here, it might cause the deadline to happen in the past, but
+  // there is resilience to this.
+  uint64_t time_until_deadline_s =
+      (time_slots_to_end * expected_dkg_timespan_) / time_slots_in_dkg_;
+
```

Comment:
Your changes have reverted the time allowed per slot back to something that depends on the number of states. The reason behind adding timer_per_slot_ is so that we can add more states in without reducing the time for each state, because the expected_dkg_timespan_ is fixed. I think this is important as the empirical values were calculated without states like the dry run and notarisations and without a fixed amount of time per slot you are eating into the time allowed for each state when you add more states.

Reply:
Oh right, I see we were coming from different directions here. I didn't get that just there and so refactored it when I noticed a possible issue with the code, which is that using doubles can result in divergent behaviour across machine architectures. I also thought it was incorrect that the DKG time was / 100 whereas the slots totalled 110, but seems I was not correct here.

Would you agree with the following change:
Change the map to be conceptually the base time a state should take
Have a lookup of a state modifer (1x usually)
?